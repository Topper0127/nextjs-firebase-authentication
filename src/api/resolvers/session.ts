import { ResolverContext } from '@typeDefs/resolver';
import { EXPIRES_IN } from '@constants/cookie';

export default {
  Mutation: {
    signIn: async (
      _: any,
      { email, password }: { email: string; password: string },
      { firebaseAdmin, firebase }: ResolverContext
    ) => {
      const {
        user,
      } = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);

      const idToken = await user?.getIdToken();
      const sessionToken = await firebaseAdmin
        .auth()
        .createSessionCookie(idToken || '', {
          expiresIn: EXPIRES_IN,
        });

      // We manage the session ourselves.
      await firebase.auth().signOut();

      return { sessionToken };
    },
    signUp: async (
      _: any,
      {
        username,
        email,
        password,
      }: { username: string; email: string; password: string },
      { firebaseAdmin, firebase }: ResolverContext
    ) => {
      await firebaseAdmin.auth().createUser({
        email,
        password,
        displayName: username,
      });

      const {
        user,
      } = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);

      const idToken = await user?.getIdToken();
      const sessionToken = await firebaseAdmin
        .auth()
        .createSessionCookie(idToken || '', {
          expiresIn: EXPIRES_IN,
        });

      // We manage the session ourselves.
      await firebase.auth().signOut();

      return { sessionToken };
    },
    passwordForgot: async (
      _: any,
      { email }: { email: string },
      { firebase }: ResolverContext
    ) => {
      // await firebaseAdmin.auth().generatePasswordResetLink(email);

      await firebase.auth().sendSignInLinkToEmail(email, {
        handleCodeInApp: true,
        url: process.env.BASE_URL || '',
      });

      return true;
    },
    passwordChange: async (
      _: any,
      { password }: { password: string },
      { me, firebaseAdmin }: ResolverContext
    ) => {
      await firebaseAdmin.auth().updateUser(me.uid, {
        password,
      });

      return true;
    },
  },
};
