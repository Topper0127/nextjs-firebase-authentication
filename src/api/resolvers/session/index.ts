import {
  ObjectType,
  Field,
  Arg,
  Ctx,
  Resolver,
  Mutation,
  UseMiddleware,
} from 'type-graphql';

import { ResolverContext } from '@typeDefs/resolver';
import { EXPIRES_IN } from '@constants/cookie';
import firebase from '@services/firebase/client';
import firebaseAdmin from '@services/firebase/admin';
import { inviteToRevue } from '@services/revue';
import { inviteToConvertkit } from '@services/convertkit';
import { isAuthenticated } from '@api/middleware/resolver/isAuthenticated';

@ObjectType()
class SessionToken {
  @Field()
  token: string;
}

@Resolver()
export default class SessionResolver {
  @Mutation(() => SessionToken)
  async signIn(
    @Arg('email') email: string,
    @Arg('password') password: string
  ): Promise<SessionToken> {
    let result;

    try {
      result = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);
    } catch (error) {
      throw new Error(error);
    }

    if (!result.user) {
      throw new Error('No user found.');
    }

    const idToken = await result.user.getIdToken();

    const token = await firebaseAdmin
      .auth()
      .createSessionCookie(idToken, {
        expiresIn: EXPIRES_IN,
      });

    if (!token) {
      throw new Error('Not able to create a session cookie.');
    }

    // We manage the session ourselves.
    await firebase.auth().signOut();

    return { token };
  }

  @Mutation(() => SessionToken)
  async signUp(
    @Arg('username') username: string,
    @Arg('email') email: string,
    @Arg('password') password: string
  ): Promise<SessionToken> {
    try {
      await firebaseAdmin.auth().createUser({
        email,
        password,
        displayName: username,
      });
    } catch (error) {
      throw new Error(error);
    }

    const { user } = await firebase
      .auth()
      .signInWithEmailAndPassword(email, password);

    if (!user) {
      throw new Error('No user found.');
    }

    const idToken = await user.getIdToken();

    const token = await firebaseAdmin
      .auth()
      .createSessionCookie(idToken, {
        expiresIn: EXPIRES_IN,
      });

    // We manage the session ourselves.
    await firebase.auth().signOut();

    try {
      inviteToConvertkit(email, username);
    } catch (error) {
      console.log(error);
    }

    try {
      inviteToRevue(email, username);
    } catch (error) {
      console.log(error);
    }

    return { token };
  }

  @Mutation(() => Boolean)
  async passwordForgot(
    @Arg('email') email: string
  ): Promise<Boolean> {
    try {
      await firebase.auth().sendPasswordResetEmail(email);
    } catch (error) {
      throw new Error(error);
    }

    return true;
  }

  @Mutation(() => Boolean)
  @UseMiddleware(isAuthenticated)
  async passwordChange(
    @Arg('password') password: string,
    @Ctx() ctx: ResolverContext
  ): Promise<Boolean> {
    try {
      await firebaseAdmin.auth().updateUser(ctx.me!.uid, {
        password,
      });
    } catch (error) {
      throw new Error(error);
    }

    return true;
  }

  @Mutation(() => Boolean)
  @UseMiddleware(isAuthenticated)
  async emailChange(
    @Arg('email') email: string,
    @Ctx() ctx: ResolverContext
  ): Promise<Boolean> {
    try {
      await firebaseAdmin.auth().updateUser(ctx.me!.uid, {
        email,
      });
    } catch (error) {
      throw new Error(error);
    }

    return true;
  }
}
