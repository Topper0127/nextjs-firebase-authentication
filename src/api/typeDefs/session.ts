import { gql } from 'apollo-server-micro';

export default gql`
  extend type Mutation {
    signIn(idToken: String!): SessionToken!

    signUp(
      username: String!
      email: String!
      password: String!
    ): Boolean

    passwordForgot(email: String!): Boolean

    passwordChange(password: String!): Boolean
  }

  type SessionToken {
    sessionToken: String!
  }
`;
