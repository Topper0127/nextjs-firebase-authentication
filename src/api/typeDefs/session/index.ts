import { gql } from 'apollo-server-micro';

export default gql`
  extend type Mutation {
    signIn(email: String!, password: String!): SessionToken!

    signUp(
      username: String!
      email: String!
      password: String!
    ): SessionToken!

    passwordForgot(email: String!): Boolean

    passwordChange(password: String!): Boolean
  }

  type SessionToken {
    sessionToken: String!
  }
`;
