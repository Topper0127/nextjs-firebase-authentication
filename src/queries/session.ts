import gql from 'graphql-tag';

export const SIGN_UP = gql`
  mutation SignUp(
    $username: String!
    $email: String!
    $password: String!
  ) {
    signUp(username: $username, email: $email, password: $password) {
      token
    }
  }
`;

export const SIGN_IN = gql`
  mutation SignIn($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      token
    }
  }
`;

export const PASSWORD_CHANGE = gql`
  mutation PasswordChange($password: String!) {
    passwordChange(password: $password)
  }
`;

export const PASSWORD_FORGOT = gql`
  mutation PasswordForgot($email: String!) {
    passwordForgot(email: $email)
  }
`;

export const EMAIL_CHANGE = gql`
  mutation EmailChange($email: String!) {
    emailChange(email: $email)
  }
`;
