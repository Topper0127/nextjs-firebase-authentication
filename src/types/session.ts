export type Session = string | null;
