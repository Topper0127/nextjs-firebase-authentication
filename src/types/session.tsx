import { AuthUser } from '@typeDefs/authUser';

export type Session = {
  authUser: AuthUser | null;
  isSessionChecked: boolean;
};
