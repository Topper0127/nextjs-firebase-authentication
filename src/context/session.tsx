import React from 'react';

import { Session } from '@typeDefs/session';

const SessionContext = React.createContext({
  authUser: null,
  isSessionChecked: false,
} as Session);

export default SessionContext;
