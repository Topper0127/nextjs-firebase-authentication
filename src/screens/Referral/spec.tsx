import renderer from 'react-test-renderer';

import { Session } from '@typeDefs/session';
import SessionContext from '@context/session';

import ReferralPage from '.';

describe('ReferralPage', () => {
  let component: any;
  let getComponent: any;
  let defaultProps: any;

  beforeEach(() => {
    defaultProps = {};

    getComponent = (props: any = defaultProps, session: Session) => (
      <SessionContext.Provider value={session}>
        <ReferralPage {...props} />
      </SessionContext.Provider>
    );
  });

  it('does not render when not authorized', () => {
    component = renderer.create(
      getComponent(null, { authUser: null, isSessionChecked: true })
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('renders when authorized', () => {
  //   component = renderer.create(
  //     getComponent(null, {
  //       authUser: { uid: '1' },
  //       isSessionChecked: true,
  //     })
  //   );

  //   const tree = component.toJSON();
  //   expect(tree).toMatchSnapshot();
  // });
});
