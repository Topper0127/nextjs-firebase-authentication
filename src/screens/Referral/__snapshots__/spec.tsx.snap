// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferralPage does not render when not authorized 1`] = `null`;
